### ISSUE TITLE
Cancelling JavaExec task should keep client alive until forked process disappears

### ISSUE THREAD

**Issue by gkastrinis** [gradle/gradle#15755]
I have a dummy project (a single groovy file) that calls an external process (via Process.start) and adds a shutdown hook (for example, for handling Ctrl+C). 

### Expected Behavior
Either there should be a way to start gradlew without a daemon, or that pipeline should handle shutdown hooks correctly.

### Current Behavior
If I run `groovy Foo.groovy` the pipeline is as expected. When I ctrl+c the hook runs.

If I run the same example via gradlew, a daemon process is spawned and ctrl+c probably just kills that process (if my understanding is correct via googling around) and the hook is never invoked.

If I first do `gradlew distZip` and then invoke the executable that way, again the hook is invoked as expected

### Steps to Reproduce 
see attached zip

### Your Environment
linux, gradle 6.5, java 8

[FOO.zip](https://github.com/gradle/gradle/files/5801700/FOO.zip)

**Comment by donat** [comment]
Gradle handles CTRL+C events since [version 4.8](https://github.com/gradle/gradle/issues/2128).

I checked your reproducer with Gradle 7.1.1. It seems like the shutdown hook is executed, only the output is not visible. If you replace the `println` statement with a file write in the shutdown hook, the file operation is executed when ctrl+c is pressed.

The problem in your case is that the Gradle command-line client shuts down too early. When ctrl+c is pressed the, the client sends a message to the daemon, that calls `Process.destroy()` on the forked process (i.e. the process created by the `JavaExec` task). That triggers the `println` in the shutdown hook and the output is sent back from the daemon to the command-line client. Since this happens asynchronously, the client already is gone when the message arrives. There's an exception in the daemon log that supports this explanation:

```
11:01:13.970 [DEBUG] [org.gradle.launcher.daemon.server.exec.ReturnResult] Daemon is dispatching the build result: Failure[value=org.gradle.initialization.ReportedException: org.gradle.internal.exceptions.LocationAwareException: Execution failed for task ':run'.]
11:01:13.970 [DEBUG] [org.gradle.launcher.daemon.server.DaemonStateCoordinator] daemon is running. Sleeping until state changes.
11:01:13.971 [DEBUG] [org.gradle.launcher.daemon.server.SynchronizedDispatchConnection] thread 122: dispatching class org.gradle.launcher.daemon.protocol.Failure
11:01:13.972 [WARN] [org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler] Unable to execute command Build{id=0cab4dce-ed78-4017-9a0c-53df48d921d3, currentDir=/Users/donat/Downloads/FOO} from socket connection from /0:0:0:0:0:0:0:1:62156 to /0:0:0:0:0:0:0:1:62216. Dispatching the failure to the daemon client
org.gradle.internal.remote.internal.MessageIOException: Could not write message Failure[value=org.gradle.initialization.ReportedException: org.gradle.internal.exceptions.LocationAwareException: Execution failed for task ':run'.] to '/0:0:0:0:0:0:0:1:62216'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.dispatch(SocketConnection.java:124)
	at org.gradle.launcher.daemon.server.SynchronizedDispatchConnection.dispatch(SynchronizedDispatchConnection.java:51)
	at org.gradle.launcher.daemon.server.DefaultDaemonConnection.completed(DefaultDaemonConnection.java:136)
	at org.gradle.launcher.daemon.server.exec.ReturnResult.execute(ReturnResult.java:47)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.api.HandleReportStatus.execute(HandleReportStatus.java:33)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.HandleCancel.execute(HandleCancel.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.api.HandleStop.execute(HandleStop.java:45)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:122)
	at org.gradle.launcher.daemon.server.exec.DaemonCommandExecuter.executeCommand(DaemonCommandExecuter.java:48)
	at org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler$ConnectionWorker.handleCommand(DefaultIncomingConnectionHandler.java:160)
	at org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler$ConnectionWorker.receiveAndHandleCommand(DefaultIncomingConnectionHandler.java:133)
	at org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler$ConnectionWorker.run(DefaultIncomingConnectionHandler.java:121)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.esotericsoftware.kryo.KryoException: java.io.IOException: Broken pipe
	at com.esotericsoftware.kryo.io.Output.flush(Output.java:156)
	at com.esotericsoftware.kryo.io.Output.require(Output.java:134)
	at com.esotericsoftware.kryo.io.Output.writeBytes(Output.java:219)
	at org.gradle.internal.serialize.kryo.KryoBackedEncoder.writeBytes(KryoBackedEncoder.java:43)
	at org.gradle.internal.serialize.AbstractEncoder$EncoderStream.write(AbstractEncoder.java:71)
	at java.base/java.io.ObjectOutputStream$BlockDataOutputStream.drain(ObjectOutputStream.java:1883)
	at java.base/java.io.ObjectOutputStream$BlockDataOutputStream.flush(ObjectOutputStream.java:1828)
	at java.base/java.io.ObjectOutputStream.flush(ObjectOutputStream.java:722)
	at org.gradle.internal.serialize.Message.send(Message.java:40)
	at org.gradle.internal.serialize.BaseSerializerFactory$ThrowableSerializer.write(BaseSerializerFactory.java:244)
	at org.gradle.internal.serialize.BaseSerializerFactory$ThrowableSerializer.write(BaseSerializerFactory.java:238)
	at org.gradle.launcher.daemon.protocol.DaemonMessageSerializer$FailureSerializer.write(DaemonMessageSerializer.java:135)
	at org.gradle.launcher.daemon.protocol.DaemonMessageSerializer$FailureSerializer.write(DaemonMessageSerializer.java:126)
	at org.gradle.internal.serialize.DefaultSerializerRegistry$TaggedTypeSerializer.write(DefaultSerializerRegistry.java:150)
	at org.gradle.internal.serialize.Serializers$StatefulSerializerAdapter$2.write(Serializers.java:46)
	at org.gradle.internal.remote.internal.inet.SocketConnection.dispatch(SocketConnection.java:116)
	... 20 more
Caused by: java.io.IOException: Broken pipe
	at java.base/sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at java.base/sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:113)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:58)
	at java.base/sun.nio.ch.IOUtil.write(IOUtil.java:50)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:466)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeWithNonBlockingRetry(SocketConnection.java:272)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeBufferToChannel(SocketConnection.java:260)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.write(SocketConnection.java:246)
	at com.esotericsoftware.kryo.io.Output.flush(Output.java:154)
	... 35 more
11:01:13.973 [DEBUG] [org.gradle.launcher.daemon.server.SynchronizedDispatchConnection] thread 122: dispatching class org.gradle.launcher.daemon.protocol.Failure
11:01:13.973 [DEBUG] [org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler] Finishing executing command: Build{id=0cab4dce-ed78-4017-9a0c-53df48d921d3, currentDir=/Users/donat/Downloads/FOO}
11:01:13.974 [DEBUG] [org.gradle.launcher.daemon.server.SynchronizedDispatchConnection] thread 122: stopping connection
11:01:13.974 [DEBUG] [org.gradle.launcher.daemon.server.SynchronizedDispatchConnection] thread 122: stopping connection
11:01:13.974 [ERROR] [org.gradle.internal.concurrent.DefaultExecutorFactory] Failed to execute org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler$ConnectionWorker@60d59dc0.
org.gradle.internal.remote.internal.MessageIOException: Could not write '/0:0:0:0:0:0:0:1:62216'.
	at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:134)
	at org.gradle.internal.remote.internal.inet.SocketConnection$1.close(SocketConnection.java:142)
	at org.gradle.internal.concurrent.CompositeStoppable$2.stop(CompositeStoppable.java:88)
	at org.gradle.internal.concurrent.CompositeStoppable.stop(CompositeStoppable.java:103)
	at org.gradle.internal.remote.internal.inet.SocketConnection.stop(SocketConnection.java:139)
	at org.gradle.launcher.daemon.server.SynchronizedDispatchConnection.stop(SynchronizedDispatchConnection.java:75)
	at org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler$ConnectionWorker.receiveAndHandleCommand(DefaultIncomingConnectionHandler.java:139)
	at org.gradle.launcher.daemon.server.DefaultIncomingConnectionHandler$ConnectionWorker.run(DefaultIncomingConnectionHandler.java:121)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.esotericsoftware.kryo.KryoException: java.nio.channels.ClosedChannelException
	at com.esotericsoftware.kryo.io.Output.flush(Output.java:156)
	at org.gradle.internal.serialize.kryo.KryoBackedEncoder.flush(KryoBackedEncoder.java:85)
	at org.gradle.internal.remote.internal.inet.SocketConnection.flush(SocketConnection.java:131)
	... 13 more
Caused by: java.nio.channels.ClosedChannelException
	at java.base/sun.nio.ch.SocketChannelImpl.ensureOpenAndConnected(SocketChannelImpl.java:180)
	at java.base/sun.nio.ch.SocketChannelImpl.beginWrite(SocketChannelImpl.java:425)
	at java.base/sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:460)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeWithNonBlockingRetry(SocketConnection.java:272)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.writeBufferToChannel(SocketConnection.java:260)
	at org.gradle.internal.remote.internal.inet.SocketConnection$SocketOutputStream.write(SocketConnection.java:246)
	at com.esotericsoftware.kryo.io.Output.flush(Output.java:154)
	... 15 more
```

**Comment by vlsi** [comment]
I believe this issue will be resolved by #18756
